yellowbrick/text/correlation.py

```diff
@@ -0,0 +1,251 @@
+from yellowbrick.exceptions import YellowbrickValueError
+
+##########################################################################
+## Word Correlation Plot Visualizer
+##########################################################################
+
+class WordCorrelationPlot(TextVisualizer):
+    """
+    Word correlation illustrates the extent to which words in a corpus appear in the
+    same documents.
+
+    WordCorrelationPlot visualizes the binary correlation between words across
+    documents as a heatmap. The correlation is defined using the phi-coefficient or
+    mean square contingency coefficient between any two words m and n.
+
+    Parameters
+    ----------
+
+
+    Attributes
+    ----------
+
+    """
+    def __init__(
+        self,
+        words,
+        ignore_case=False,
+        ngram_range=(1, 1),
+        ax=None,
+        cmap="RdYlBu",
+        colorbar=True,
+        fontsize=None,
+        **kwargs
+    ):
+        super(WordCorrelationPlot, self).__init__(ax=ax, **kwargs)
+
+        # Visual parameters
+        self.fontsize = fontsize
+        self.colorbar = colorbar
+        self.cmap = color_sequence(cmap)
+
+        # Fitting parameters
+        self.ignore_case = ignore_case
+        self.ngram_range = ngram_range
+        self.words = self._construct_terms(words, ignore_case)
+
+    def _construct_terms(self, words, ignore_case):
+        """
+        Constructs the list of terms to be plotted based on the provided words.
+        """
+        # Remove surrounding whitespace
+        terms = [word.strip() for word in words if len(word.strip()) > 0]
+
+        if len(terms) == 0:
+            raise YellowbrickValueError("Must provide at least one word to plot.")
+
+        # Convert to lowercase if ignore_case is set
+        if ignore_case:
+            terms = [word.lower() for word in terms]
+
+        # Sort and remove duplicates
+        return sorted(set(terms))
+
+    def _compute_coefficient(self, m, n):
+        """
+        Computes the phi-coefficient for two words m and n.
+        """
+        m_col = self.doc_term_matrix_.getcol(self.term_index_[m])
+        n_col = self.doc_term_matrix_.getcol(self.term_index_[n])
+        both = m_col.multiply(n_col).sum()
+        m_total = m_col.sum()
+        n_total = n_col.sum()
+        only_m = m_total - both
+        only_n = n_total - both
+        neither = self.docs_ - both - only_m - only_n
+        return ((both * neither) - (only_m * only_n)) / np.sqrt(m_total * n_total * (self.docs_ - m_total) * (self.docs_ - n_total))
+
+    def _construct_term_index(self):
+        """
+        Constructs a dictionary mapping terms to indices in the sparse doc-term matrix.
+        """
+        self.term_index_ = {}
+        for i, term in enumerate(self.vocab_):
+            if term in self.words:
+                if term in self.term_index_:
+                    raise ModelError("Duplicate term in vocabulary: {}".format(term))
```

Comment:
Awesome error handling here, Could you write a test to raise it?    Also, do you think that is a different type of error than a ModelError?  Just wondering ❓

Reply:
I agree @lwgray this might be a good place to introduce a subclass of `ModelError` specifically for text.

Reply:
@lwgray @bbengfort I'm not sure I can trigger this error from the API, since sklearn _should_ not be producing a vocab with repeated words, so I would have to make the unit test more specific to `_construct_term_index`. This leads me to the question, should we assume that sklearn will behave this way? Instead of catching the error, we could convert the vocab to a `set` to ensure that the values are unique. This would protect us a bit but we might also want to know if the sklearn API changes. What do you guys think?

Reply:
@bbengfort Can you chime in?

Reply:
The sklearn API does change quite a bit, so being defensive about it is usually a good idea. Just out of curiosity, the vocab comes from `vec.get_feature_names_out()` right? Why not use `vec. vocabulary_.keys()`? This variable is a dictionary, so would guarantee uniqueness and would prevent the need for an intermediate `set`. 

If triggering this error is too difficult for a test we should put a comment above the `raise` to indicate what we thinking is going on (e.g. unexpected duplicate vocab possibly due to sklearn API change).